### ISSUE TITLE
Support for declarative project configuration (setup.cfg / pyproject.toml)

### ISSUE THREAD

**Issue by schlamar** [python-cffi/cffi#55]
There was a ticket in heptapod including a patch to support  configuration of cffi modules in setup.cfg https://foss.heptapod.net/pypy/cffi/-/issues/396

Personally, I would prefer a configuration in pyproject.toml as this seems to be the new standard everyone is adopting.

**Comment by arigo** [comment]
I won't work on it as long as the existing solution works fine, but I'm not against it, if the new way works in older systems too.  Pull requests welcome.  See e.g. https://github.com/python-cffi/cffi/actions/runs/7709709848 for the list of supported versions of Pythons.  (Ideally it should also work if the user uses some ``setuptools`` that was installed at the time of that Python release and not updated since.  I'm pretty sure that's a completely unrealistic expectation, though, so take that with a grain of salt.)

**Comment by genotrance** [comment]
Here's my setup.py:
```python
from setuptools import setup

setup(
    cffi_modules=["mcurl/gen.py:ffibuilder"]
)
```
Everything else is in pyproject.toml and it works on Python 3.6+ on Windows, Linux (glibc/musl, x86_64/i686/aarch64) and OSX. Will be nice to move this into pyproject.toml.

**Comment by JamesTheAwesomeDude** [comment]
> Everything else is in pyproject.toml and it works on Python 3.6+ on Windows, Linux (glibc/musl, x86_64/i686/aarch64) and OSX. Will be nice to move this into pyproject.toml.

@genotrance Does your project support [ABI3](https://docs.python.org/3/c-api/stable.html#stable-abi)?

In my own project, I've got that file still existing for *two* reasons: one of them is to keep `cffi_modules` referenced, and the other is [to make ABI3 work with CFFI](https://github.com/JamesTheAwesomeDude/pypqc/blob/782421f3a241e492c7d0b768853aae593100c413/setup.py#L49).

I'm not sure my solution is a good one, though, and I'm also trying to ditch `setup.py` and move to a "pure" pyproject.toml, so I'm curious how other people are handling this.

**Comment by genotrance** [comment]
Yes it supports ABI3 - [here's](https://github.com/genotrance/mcurl/blob/main/build.sh#L52) how I am specifying it via the CLI. The resulting [wheels](https://pypi.org/project/pymcurl/#files) run on every version of Python I've tested.

**Comment by genotrance** [comment]
I now do the following in `pyproject.toml` for ABI3 wheels.
```
[tool.distutils.bdist_wheel]
py_limited_api = "cp32"
```
So everything but `cffi_modules` is in `pyproject.toml`. I am able to get everything working with `cibuildwheel` which has made life so much easier but wish the ABI3 info was easier to figure out.

**Comment by mattip** [comment]
I think #241 fixes this, right? Closing. Please reopen if I am mistaken.
